### ISSUE TITLE
libtorch: unknown error getNumGPUs

### ISSUE THREAD

**Issue by bgobbi** [pytorch/pytorch#14561]
## 🐛 Bug

Hi,
I am trying to modify example.app to check for the availability of a GPU and use it if available.

Here is my code
```
#include <torch/torch.h>
#include <iostream>

int main() {
   torch::DeviceType device_type;
   if (torch::cuda::is_available()) {
     std::cout << "CUDA available! Training on GPU" << std::endl;
     device_type = torch::kCUDA;
   } else {
     std::cout << "Training on CPU" << std::endl;
     device_type = torch::kCPU;
   }
   torch::Device device(device_type);

  torch::Tensor tensor = torch::rand({2, 3}).to(device);
  std::cout << tensor << std::endl;
}
```

However if I compile and execute it (using cmake/make) I get the following error:
```
terminate called after throwing an instance of 'c10::Error'
  what():  CUDA error (30): unknown error (getNumGPUs at /pytorch/aten/src/ATen/cuda/detail/CUDAHooks.cpp:202)
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x2aaab43d7d31 in /home/myuser/dev/libtorch/libtorch/lib/libc10.so)
frame #1: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x2a (0x2aaab43d75fa in /home/myuser/dev/libtorch/libtorch/lib/libc10.so)
frame #2: at::cuda::detail::CUDAHooks::getNumGPUs() const + 0xb2 (0x2aaab5b3b582 in /home/myuser/dev/libtorch/libtorch/lib/libcaffe2_gpu.so)
frame #3: torch::cuda::device_count() + 0x35 (0x2aaaab493695 in /home/myuser/dev/libtorch/libtorch/lib/libtorch.so.1)
frame #4: torch::cuda::is_available() + 0x9 (0x2aaaab4936a9 in /home/myuser/dev/libtorch/libtorch/lib/libtorch.so.1)
frame #5: main + 0xe (0x40c9f7 in example-app)
frame #6: __libc_start_main + 0xf5 (0x2aaab460ab35 in /usr/lib64/libc.so.6)
frame #7: example-app() [0x40c049]

Abort (core dumped)
```

When run on a server with a GPU this works fine.
When removing all but this two lines the code runs fine too:
```
  torch::Tensor tensor = torch::rand({2, 3})
  std::cout << tensor << std::endl;
```
Any ideas how to fix this?
Thanks!

## Environment
I am using:  https://download.pytorch.org/libtorch/nightly/cu90/libtorch-shared-with-deps-latest.zip
downloaded yesterday.

```
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: CentOS Linux release 7.3.1611 (Core)
GCC version: (GCC) 6.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 9.0.176
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy (1.13.3)
[pip] numpydoc (0.7.0)
[pip] torch (0.4.1)
[conda] Could not collect
```

**Comment by ShahriarSS** [comment]
Hi. 
I'm having the same problem here

terminate called after throwing an instance of 'c10::Error'
  what():  CUDA error (30): unknown error (getNumGPUs at /home/shahriar/Videos/pytorch/aten/src/ATen/cuda/detail/CUDAHooks.cpp:164)
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x46 (0x7fa8fa08b956 in /usr/local/lib/libc10.so)
frame #1: at::cuda::detail::CUDAHooks::getNumGPUs() const + 0xac (0x7fa8fcdd59cc in /usr/local/lib/libcaffe2_gpu.so)
frame #2: ./testPytorch() [0x40aa50]
frame #3: main + 0x23 (0x40ae89 in ./testPytorch)
frame #4: __libc_start_main + 0xf3 (0x7fa8d8160413 in /usr/lib64/libc.so.6)
frame #5: ./testPytorch() [0x40a7de]

I tried it with both libTorch and compiling from source but the problem persisted. 
OS: Fedora 29
CUDA: V9.2.148

**Comment by ShahriarSS** [comment]
> Something like
> 
> ```
> diff --git a/torch/csrc/api/src/cuda.cpp b/torch/csrc/api/src/cuda.cpp
> index a830e6572..45dcbf3bf 100644
> --- a/torch/csrc/api/src/cuda.cpp
> +++ b/torch/csrc/api/src/cuda.cpp
> @@ -7,6 +7,7 @@
>  namespace torch {
>  namespace cuda {
>  size_t device_count() {
> +  if (!at::hasCUDA()) return 0;
>    return at::globalContext().getNumGPUs();
>  }
>  
> @@ -14,7 +15,7 @@ bool is_available() {
>    // NB: the semantics of this are different from at::globalContext().hasCUDA();
>    // ATen's function tells you if you have a working driver and CUDA build,
>    // whereas this function also tells you if you actually have any GPUs.
> -  return cuda::device_count() > 0;
> +  return at::hasCUDA() && cuda::device_count() > 0;
>  }
>  
>  bool cudnn_is_available() {
> ```

I don't know if I did it right but I did these and compiled pytorch from source but the problem is still there. when I cout at::hasCUDA() the output is 1.

**Comment by ShahriarSS** [comment]
It looks like the problem wasn't with Pytorch. I was just missing nvidia-driver-cuda package.

**Comment by ezyang** [comment]
Fixed more durably in #18445
